src/System.Linq/src/System/Linq/Enumerable.SelectWithCountIterator.cs

```diff
@@ -0,0 +1,62 @@
+﻿
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        internal abstract class SelectWithCountIterator<TSource> : SelectIterator<TSource>, ICollection<TSource>
+        {
+            public override abstract Iterator<TSource> Clone();
+
+            public override abstract bool MoveNext();
+
+            public override abstract IEnumerable<TSource> Where(Func<TSource, bool> predicate);
+
+            public override abstract IEnumerable<TResult> SelectImpl<TResult>(Func<TSource, TResult> selector);
+
+            #region ICollection<TSource> implementation
+
+            public abstract int Count { get; }
+
+            public bool IsReadOnly { get { return true; } }
+
+            public void Add(TSource item)
+            {
+                throw NotImplemented.ByDesign;
+            }
+
+            public void Clear()
+            {
+                throw NotImplemented.ByDesign;
+            }
+
+            public bool Contains(TSource item)
+            {
+                throw NotImplemented.ByDesign;
```

Comment:
Yes, I meant it to be only implemented by LINQ iterators, when it makes sense (`Select`, `OrderBy`, `Reverse` iterators etc. based on array, list or `ICollection<T>` would implement it).

`CopyTo` is a little misleading here I think, because we're not copying anything - some sort of logic has to be applied to make the new collection, either a projection or change of order. Because of that, and because it would be a new interface I would name it differently (`Initialize` comes to my mind, but there might be a better one). 

Recreating the logic within *System.Linq* instead of pushing it to *System.Collections.Generic* seems reasonable. It would safe us trouble with `InternalsVisibleTo` and would make it more contained within just one library. 
bly name it differently. Recreating the logic within *System.Linq* instead of pushing it to *System.Collections.Generic* seems reasonable. It would safe us trouble with `InternalsVisibleTo` and would make it more contained within just one library. It would just call `List<T>` with `capacity` parameter and than pass it to `Initialize` method to fill with data.